src/unzip.ts

```diff
@@ -0,0 +1,125 @@
+import * as fs from "fs";
+import * as path from "path";
+import * as zlib from "zlib";
+import { Readable, Transform, TransformCallback } from "stream";
+import { promisify } from "util";
+import { FirebaseError } from "./error";
+import { pipeline } from "stream";
+
+const pipelineAsync = promisify(pipeline);
+
+interface ZipEntry {
+  compressedSize: number;
+  uncompressedSize: number;
+  fileNameLength: number;
+  extraLength: number;
+  fileName: string;
+  headerSize: number;
+  compressedData: Buffer;
+}
+
+const readUInt32LE = (buf: Buffer, offset: number): number => {
+  return (
+    (buf[offset] | (buf[offset + 1] << 8) | (buf[offset + 2] << 16) | (buf[offset + 3] << 24)) >>> 0
+  );
+};
+
+const extractEntriesFromBuffer = async (data: Buffer, outputDir: string): Promise<void> => {
+  let position = 0;
+
+  while (position < data.length) {
+    const entryHeader = data.slice(position, position + 30);
+    const entry: ZipEntry = {} as ZipEntry;
+
+    if (readUInt32LE(entryHeader, 0) !== 0x04034b50) {
+      break;
+    }
+
+    entry.compressedSize = readUInt32LE(entryHeader, 18);
+    entry.uncompressedSize = readUInt32LE(entryHeader, 22);
+    entry.fileNameLength = entryHeader.readUInt16LE(26);
+    entry.extraLength = entryHeader.readUInt16LE(28);
+    entry.fileName = data.toString("utf-8", position + 30, position + 30 + entry.fileNameLength);
+    entry.headerSize = 30 + entry.fileNameLength + entry.extraLength;
+    entry.compressedData = data.slice(
+      position + entry.headerSize,
+      position + entry.headerSize + entry.compressedSize
+    );
+
+    entry.fileName = entry.fileName.replace(/\//g, path.sep);
+
+    const outputFilePath = path.normalize(path.join(outputDir, entry.fileName));
+
+    console.log(`Processing entry: \${entry.fileName}`);
+    if (entry.fileName.endsWith(path.sep)) {
+      await fs.promises.mkdir(outputFilePath, { recursive: true });
+    } else {
+      const parentDir = outputFilePath.substring(0, outputFilePath.lastIndexOf(path.sep));
+      await fs.promises.mkdir(parentDir, { recursive: true });
+
+      const compressionMethod = entryHeader.readUInt16LE(8);
+      if (compressionMethod === 0) {
+        // Store (no compression)
+        console.log(`Writing file: \${outputFilePath}`);
```

Comment:
Please switch this to logger.debug(). That way, the logs will only appear in --debug mode.